Add zombie wall breaching

Caaz opened the pull request:
[Allows zombies to break through walls](https://youtu.be/2pogVWh6OIk)! Configurable, of course, by default it's off.

In general, this is an attempt to make [my tactics](https://github.com/Zombiefied7/ZombiefiedZombieApocalpyse/issues/4) useless

How does this work?
- Given `allowBreaching` is `true`
  - Add a Toil to the ZombieHunt job to get the blocking building on a path to the pawn
  - When a blocking building is found, target it add a toil to move to that target.
  - Continue the job as normal, `FollowAndMeleeAttack` won't follow buildings, but it'll melee attack it 🤔 
  
What else is done
- Some general changes to how the jobgiver assigns a target for hunting.
  - Some formatting changes to reduce indentation
  - Some renaming to make variables and methods a little more clear
  - Considers pawns behind a wall further away than pawns right next to them, meaning in general they'll path towards a near pawn first.
    - Only does this check if we can path through walls anyway.

Commit:
Check whole map when finding targets for zombies!

Also, add a class method to zombies for pathing
General clean up of the JobGiver for ZombieResponse.

Commit:
Additional debugging

Commit:
This don't do the do

Commit:
Allow zombies to break down walls

Commit:
Explain some things

Commit:
Add zombie wall breaching!

Potentially break other things!

Commit:
revert range

Zombiefied7 commented on Source/JobGiver_ZombieResponse.cs:
```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
+                // if predator can't reach the prey, skip.
+                if (!predator.CanReach(prey, ZombiefiedMod.allowBreaching)) continue;
+                // if predator can reach prey, but it has to go through walls, add to distance, to ensure un-blocked pawns are considered closer, and thus easier to get to
+                if (ZombiefiedMod.allowBreaching && !predator.CanReach(prey, false)) distance *= 5;
```
You dont want to call CanReach twice if avoidable because pathing is expensive

Caaz replied:
Hmm. Got an idea for how to do it? In this case we only do it twice if breaching is allowed, we have to do it twice to figure out if it can reach it without going through walls, and with going through walls.

Zombiefied7 replied:
Probably just check for distance and if for some reason it cant reach the prey eventho it can destroy anything in its path just let the job fail

Caaz replied:
That should be doable. I guess this check is only to consider pawns behind walls harder to get to, to get the zombie to get something without trying to breach all the time.

What do you think about having this check behind a configuration just in case? Default it to off, people can turn it on if they want to, warn em about a performance increase because if it or something

Zombiefied7 replied:
I dont mind but then noone is ever gonna use it. People dont ususally check options. Also theyre gonna enable it thinking they wouldnt care about the performance hit and then theyre still gonna hate the performance hit

Zombiefied7 replied:
But I also think this can work out well enough to be included by default just needs some work: Performance needs to be good, there needs to be a reason for them to attack through walls (maybe embrasures or sounds from inside the walls).
Right now they just have wallhack as you called it. that is kinda weird but has potential to be neat

Zombiefied7 replied:
> That should be doable. I guess this check is only to consider pawns behind walls harder to get to, to get the zombie to get something without trying to breach all the time.

```
if(!predator.CanReach(prey, false))
{
   if(ZombiefiedMod.allowBreaching)
   {
      distance *=5;
   } else {
      continue;
   }
}
```

Caaz replied:
>there needs to be a reason for them to attack through walls (maybe embrasures or sounds from inside the walls)

Totally want to work out some reasoning for it yeah. I want to look into the fog of war mod and figure out what they did to figure out which buildings are able to be seen through. Embrasures are, for example, and I imagine mods that add glass probably are too. I figure zombies wanting to path through those types of things makes sense.

Right now though, it's terrifying, and certainly an increase in difficulty.

Zombiefied7 replied:
Might also just check if embrasures exist in the game (so an embrasures mod is active) and enable allowBreaching if that is the case

Zombiefied7 replied:
Also might allowBreaching only when moving towards a noisy location, not when hunting.

Just ideas btw

Zombiefied7 replied:
Or maybe allowBreaching only when moving towards a noisy location and then they also find a prey behind walls

Caaz replied:
>``` 
> if(!predator.CanReach(prey, false))
> {
>    if(ZombiefiedMod.allowBreaching)
>    {
>       distance *=5;
>    } else {
>       continue;
>    }
> }```

Ooooh wait no I get it, so if the option is on, consider the distance higher regardless... But that means the job can potentially try to target a pawn that isn't pathable through other means, like weird terrain. The check right now should at least ensure it can destroy things on the path, but if it's like, super deep water or something weird that'd be a problem. I guess it'd raise PatherFailed in the job that could be handled though.

> Or maybe allowBreaching only when moving towards a noisy location and then they also find a prey behind walls

Oooh. Hmm. well, they already try to get to noisy locations, so that part is handled, and I've noticed they tend to gather at walls. Maybe we can just have breaching only happen when they're nearby already? 🤔Sounds super doable, would just require putting a distance check on the blocking buildings toil

I think if it's on when doing noisy locations though, I would think that'd mean they'd more-regularly bust down walls if you have a wall around your houses. At least with the hunting thing they've gotta be in your general vicinity.

Zombiefied7 commented on Source/JobGiver_ZombieResponse.cs:
```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
         {
-            //List<Pawn> allPawnsSpawned = predator.Map.mapPawns.AllPawnsSpawned;
-            List<Thing> allThingsRegion = predator.GetRegion().ListerThings.AllThings;
+            List<Thing> allThingsRegion = predator.Map.listerThings.AllThings;
```
Checking the whole map is expensive. Is region not enough?

Caaz replied:
You know what I did that to test, but it might not be enough if region limits it to just the room it's in, 🤔 Can't target a pawn through a wall, and makes the feature just not work. 

However, we can probably at the very least put this behind a check for the feature configuration. I'll test and update that.

Zombiefied7 replied:
Ah so region refers to room? I didnt know. Just keep in mind that ZombieResponse gets called virtually always and for every single zombie so performance is important here and iirc using region instead of map increased performance by a lot

Caaz replied:
It could, I feel like that's what I was running into when initially testing, but I could be wrong on that.

Caaz replied:
Ah maybe I'm a bit wrong on that, so regions are kind of like rooms but more partitions of the map, the problem is they partition off where rooms intersect.  The problem with this, is I can't get pawns on the other side of a wall if this is the case.

https://www.dropbox.com/s/a1w6jaxty9hzobv/Screenshot%20%2891%29.png?dl=0

Was testing with this as the line to hopefully solve it using the `RegionType` parameter, but it doesn't seem to work here.
```cs
List<Thing> allThingsRegion = predator.GetRegion(RegionType.Set_All).ListerThings.ThingsMatching(ThingRequest.ForGroup(ThingRequestGroup.Pawn));
``` 

Edit: After messing around I'm realizing because the `GetRegion` method will return the region the pawn is standing on, sooooo it's always gonna be partitioned. Hmmm. By adding the RegionType it basically adds a validation to say "if the region isn't this type, don't return it"

Zombiefied7 commented on Source/JobDriver_ZombieHunt.cs:
```diff
@@ -10,29 +10,51 @@ namespace Zombiefied
-        public override void ExposeData()
-        {
-            base.ExposeData();
-            Scribe_Values.Look<int>(ref this.numMeleeAttacksMade, "numMeleeAttacksMade", 0, false);
```
Is there a reason to remove this? iirc this was to fix some bug

Caaz replied:
Probably something dumb I was testing, I'll add it back cause I can't think of why it was removed. good catch

Zombiefied7 commented on Source/JobDriver_ZombieHunt.cs:
```diff
@@ -10,29 +10,51 @@ namespace Zombiefied
-                    if (this.pawn.CurJob == null || this.pawn.jobs.curDriver != this)
-                    {
-                        return;
-                    }
```
Is there a reason to remove this

Caaz replied:
oof, will revert, good catch, was learning things.

Zombiefied7 commented on Source/JobGiver_ZombieResponse.cs:
```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
+            {
+                Pawn prey = item as Pawn;
+
+                // if prey is a pawn or the predator, skip.
```
// if prey is **not** a pawn or the predator, skip.

Zombiefied7 commented on Source/JobDriver_ZombieHunt.cs:
```diff
@@ -43,31 +65,6 @@ protected override IEnumerable<Toil> MakeNewToils()
             }).FailOnDespawnedOrNull(TargetIndex.A);
         }
 
-        public override void Notify_PatherFailed()
```
I dont see how they still attack doors after removing this. let me know pls

Zombiefied7 replied:
I guess JobDriver already implements this? It is working but I dont understand why

Caaz replied:
🤔 I didn't test it, but I will certainly re-add the door feature if this broke it, now that I understand how it works.

I think I'll have to re-work it a little bit but that shouldn't be too bad.

Zombiefied7 replied:
It still works thats why Im confused. You removed it but it still works *shrugs*

Zombiefied7 replied:
This functionality is copied from manhunting animals so Im not completely sure what is going on. I assume this block of code is only for some specific scenarios so it is not immediately noticeable. So I would just keep it

Caaz replied:
Well, by default the `FollowAndMeleeAttack` toil uses the pawn's pather and that uses the mode that allows passing through doors, so it doesn't consider it a pather failure when the door is in the way. Manhunting animals might use a different pather by default that doesn't consider doors okay. 🤔

Zombiefied7 commented on Source/JobGiver_ZombieResponse.cs:
```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
         }
 
-        public bool IsAcceptablePreyFor(Pawn predator, Pawn prey, float distance)
+        public bool IsAcceptablePreyFor(Pawn predator, Pawn prey)
```
these are clean af

Zombiefied7 commented on Source/JobGiver_ZombieResponse.cs:
```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
-                return new Job(ZombiefiedMod.zombieHunt, thing)
-                {
-                    killIncappedTarget = true,
-                    expiryInterval = (int)(Rand.RangeSeeded(1f, 2f, Find.TickManager.TicksAbs) * 700),
```
expiryInterval is important because without it zombies will follow a pawn forever and never lose sight and they shouldnt do that

Caaz replied:
oh man I just stumbled upon the usage in jobs in the decompilation, no idea why it wasn't showing up before. blaaah. Anyway I went through and tested it and yeah definitely important. Gonna add it in.

Commit:
Add back expiryInterval

I'm a big dumb, it is important

Commit:
Fix comment

Commit:
Add back in these functions and such

Probably no real reason they were removed.

Commit:
Use built in distance calculation, and maxNumMeleeAttacks

Commit:
Use breachChance rather than allowBreaching

This adjusts gameplay difficulty, generally making this feature way more playable.

Commit:
Use blocker pathing only when normal pathing fails!

This'll mean when there's an open path to a pawn, we try to take the non-wall-busting path first!

Commit:
Remove commented code.

Caaz commented:
Alright, a lot of functional changes since the last check in on this.
- Realized my dumbness in the expiryInterval stuff
  - For whatever reason, dnSpy search wasn't showing the functionality of expiryInterval, but I wound up stumbling upon the actual use in decompilation and I get it now, it's not deprecated, I'm just a dumb.
- Addressed comment fixes.
- Added back things deleted without real reason
  - For sure a bruteforce way of learning what things do to see what happens when they stop doing them
- Using the built in distance function, which is probably functionally normal, but reduces the need to think about it.
- Re-add-in maxNumMeleeAttacks, cause without it zombies attack until they break a wall, and that's too spooky.
- Instead of a boolean for this feature, use a `breakChance` integer (which winds up being a float in the code, essentially). That chance will decide whether or not a pawn is applicable to be targeted by the jobGiver. Since the job itself can handle either case, it doesn't need to use the variable at all.
  - Because of how often the jobgiver tries to give jobs, this feels like even really, really low values are high. To combat this, I've used it with the distance calculation, to reduce the chance the further a zombie is away. Meaning, even at 100% chance of breach chance, at the max distance (defined by the job itself) of 7f means there should only be a 7%ish chance of targeting. a pawn.
  - Effectively, this means if you don't want zombies to attack your walls often, don't stay near the walls.
  
Going to be playing with it on my normal save for a while to see if anything pops up but this miiiiight be playable.
 
Edit: 
- Tried to use the blocker targeting toil only when the normal pather for FollowAndMeleeAttack fails. to ensure the pather tries a non-breaching path first, but that failed so instead now it's checking when the job is created whether or not a simple path is possible.

Commit:
Fix job

Okay now, we should path not through walls if a free path is available.

Commit:
Okay but do it right this time

Zombiefied7 commented:
Hey. Im busy with writing a thesis right now and dont have the capacity to check what youre doing. Anyways I do appreciate your work on this so if you like to keep at it, I will check it when Im done with the thesis

Zombiefied7 commented:
Or you can just make any changes you like and mess it all up thats fine too. We can still fix it later and its good for learning. Let me know what are your plans

Caaz commented:
Ahh, awesome good luck on that, sounds hella fancy.

I started working on my own mod as well, so I've been focused on that as well as a project I've been invited to. I haven't been in zombie mode in a little while now, but I've definitely learned a lot of little nifty stuff along the way while working on my typhon mod.

I could certainly get back to this at sometime, but it's not my focus at the moment. I'm a bit spread thin here as well lol
